aten/src/ATen/core/type.cpp

```diff
@@ -202,9 +202,17 @@ c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
     return static_cast<TypePtr>(TupleType::create(elements));
   }
 
+  if (t1->cast<FutureType>() && t2->cast<FutureType>()) {
```

Comment:
perhaps I'm missing something: why are Futures exempt from the same problem that we had with dicts? e.g. isn't it wrong for `Future[int]` and `Future[Optional[int]]` to unify into `Future[int]`?

Reply:
Future is an immutable type, there's no way to mutate what it's contained element is.

Reply:
ah, ok I buy it.

Reply:
However I tried an example and i got an assert failure from elsewhere, so it looks like we currently enforce that Future is waited on in the block it is created:
```
def int_fn():
    return 1

def none_fn():
    return None

@torch.jit.script
def test(cond):
    if cond:
        x = torch.jit._fork(int_fn)
    else:
        x = torch.jit._fork(none_fn)
    return torch.jit._wait(x)
```

> RuntimeError: fork_graph_context->uses().size() == 1 INTERNAL ASSERT FAILED at ../torch/csrc/jit/passes/inline_forked_closures.cpp:43, please report a bug to PyTorch.  (inlineForkedClosure at ../tor

So i can remove this if requested.